### ISSUE TITLE
zoneminder.sh could be better in regards to handling submodules, I rewrote it.

### ISSUE THREAD

**Issue by dekeonus** [nabbi/oubliette-ebuild-verbump#1]
The zoneminder.sh looks to be a quick and dirty hack to get the job done. It doesn't do the right thing in regards to the git submodules: The zoneminder ebuilds are for 1.36 releases they have the RtspServer pegged at a specific tag, the revbump script just pulls master.

I've re-written the script to use **jq** _app-misc/jq_ to parse output from curl interfacing with github's api.

As the script places the release's github submodule hash into the MY_SUBMODULE variables, the 9999 ebuild needs to be rewritten to handle the URIs with commit hashes rather than named refs.
I've re-written an ebuild 1.36.9999 that uses these commit hashes AND sets EGIT_BRANCH="release-1.36" so that someone wanting the 1.36 series live master can do so.
The revbump script is looking for this 1.36.9999 ebuild, but would be trivial to set it back to the 9999 ebuild

[zoneminder.sh.gz](https://github.com/nabbi/oubliette-ebuild-verbump/files/9275995/zoneminder.sh.gz)
[zoneminder-1.36.9999.ebuild.gz](https://github.com/nabbi/oubliette-ebuild-verbump/files/9275996/zoneminder-1.36.9999.ebuild.gz)
.

**Comment by dekeonus** [comment]
```
#!/bin/bash

#git repo and package subdir
repo=~/oubliette-overlay/www-misc/zoneminder
pkg="zoneminder"

zm_git_base="https://api.github.com/repos/ZoneMinder/zoneminder"
curl_cmd="curl --header 'Accept: application/vnd.github+json' --silent"
jq_cmd="jq --raw-output"

# zm.com does not always reflect latest github release
#version=$(curl --silent https://update.zoneminder.com/version.txt)
#version=$(curl --silent https://api.github.com/repos/ZoneMinder/zoneminder/releases/latest | grep tag_name | sed "s/,//" | sed 's/\s*"tag_name":\s*//' | sed 's/"//g')
#version=$(curl --silent https://api.github.com/repos/ZoneMinder/zoneminder/releases/latest | awk '/tag_name/ {split($2, arry, "\""); print arry[2]}')

function vars_by_jq() {
    local tag_sha commit_sha
    # grab latest release tag (version)
    version=$(${curl_cmd} "${zm_git_base}/releases/latest" |${jq_cmd} '.tag_name')
    # get version major.minor
    zm_maj_min=$(echo ${version}|grep -Eo '[0-9]+\.[0-9]+')
    # using release tag grab tag sha
    tag_sha=$(${curl_cmd} "${zm_git_base}/git/ref/tags/${version}" |${jq_cmd} '.object.sha')
    # using tag sha grab commit sha
    commit_sha=$(${curl_cmd} "${zm_git_base}/git/tags/${tag_sha}" |${jq_cmd} '.object.sha')
    # grab info for sub-modules using commit sha
    Crud_info=$(${curl_cmd} "${zm_git_base}/contents/web/api/app/Plugin/Crud?ref=${commit_sha}")
    CakePHPEB_info=$(${curl_cmd} "${zm_git_base}/contents/web/api/app/Plugin/CakePHP-Enum-Behavior?ref=${commit_sha}")
    RtspSrv_info=$(${curl_cmd} "${zm_git_base}/contents/dep/RtspServer?ref=${commit_sha}")
    # test and get sha for submodules
    if [ "$(echo ${Crud_info}|${jq_cmd} '.type')" != "submodule" ]; then
        # upstream changed layout or dependancies
        echo "upstream changes to Crud submodule, bailing out"
        exit 1
    else
        Crud_sha=$(echo ${Crud_info}|${jq_cmd} '.sha')
	if [ "${Crud_sha}" == "" -o "${Crud_sha}" == "null" ]; then
            # issue with github api?
            echo "could not get sha for Crud submodule, api issue?"
            exit 1
        fi
    fi
    if [ "$(echo ${CakePHPEB_info}|${jq_cmd} '.type')" != "submodule" ]; then
        # upstream changed layout or dependancies
        echo "upstream changes to CakePHP-Enum-Behavior submodule, bailing out"
        exit 1
    else
        CakePHPEB_sha=$(echo ${CakePHPEB_info}|${jq_cmd} '.sha')
	if [ "${CakePHPEB_sha}" == "" -o "${CakePHPEB_sha}" == "null" ]; then
            # issue with github api?
            echo "could not get sha for CakePHP-Enum-Behavior submodule, api issue?"
            exit 1
        fi
    fi
    if [ "$(echo ${RtspSrv_info}|${jq_cmd} '.type')" != "submodule" ]; then
        # upstream changed layout or dependancies
        echo "upstream changes to RtspServer submodule, bailing out"
        exit 1
    else
        RtspSrv_sha=$(echo ${RtspSrv_info}|${jq_cmd} '.sha')
	if [ "${RtspSrv_sha}" == "" -o "${RtspSrv_sha}" == "null" ]; then
            # issue with github api?
            echo "could not get sha for RtspServer submodule, api issue?"
            exit 1
        fi
    fi
}

function update_ebuild() {
    if [ -e ${ebuild} ] ; then
        echo "no changes needed $pkg $version ebuild exists"
        exit
    fi

    cp -v ${from} ${ebuild}
    ## I had wanted to keep trailing comments (if any) but it was too fragile
    sed -i -E -e "s/^(MY_CRUD_V=).*/\1\"$Crud_sha\"/" \
        -e "s/^(MY_CAKEPHP_V=).*/\1\"$CakePHPEB_sha\"/" \
        -e "s/^(MY_RTSP_V=).*/\1\"$RtspSrv_sha\"/" \
        ${ebuild}

#    ebuild ${ebuild} digest #depreciated
    ebuild ${ebuild} manifest
}

function push_to_overlay() {
    cd ${repo}
    git pull

    update_ebuild

    git add .
    git commit -asm "${pkg} auto-verbump"
    git push
}

vars_by_jq

ebuild="${pkg}-${version}.ebuild"
if [ "${zm_maj_min}" == "1.36" ]; then
    from="${pkg}-${zm_maj_min}.9999.ebuild"
else
    from="${pkg}-9999.ebuild"
fi

push_to_overlay

```

**Comment by nabbi** [comment]
@dekeonus I appreciate the time you've invested on this project. Yup, this wasn't meant to be elegant :smiley:  It may be a few before I have cycles to address this, so I would happily accept PR requests (and for ebuild overlay too) in accelerating merging your proposed changes. Thank you!

**Comment by dekeonus** [comment]
@nabbi cheers, has been my intent to submit PR, but until earlier this week I had no idea how to use git.
I've been reading the gentoo devmanual and related polices (GLEP66 GLEP63) regarding submitting PRs.

I'm hoping to get a PR on this (revbump script) mid weekend (UTC+10).

One of my major changes for the zoneminder ebuild will be enabling selection of apache2 or nginx (it's a lot of small changes and to multiple files).

**Comment by nabbi** [comment]
These repos fall outside of official gentoo control, so there's no formal process or policies to be overly concerned about. ie no bugzilla id etc needed here.

This might help you get started:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

Essentially; Fork these repos in github, clone your new private copy of them, make changes, add and commit them, push back to your repo fork. Then github webui will have a contribute button to PR your changes 

It might be a week or two before I can circle back with you on these, they are all good finds. And I should have more time then to look at then... so until then I added you as a collaborator into [oubliette-overlay-dev ](https://github.com/nabbi/oubliette-overlay-dev) if you need some space to test your PR process and ebuild changes while I'm out of pocket. That overlay isn't in layman but has instructions for adding it manually. I typically sandbox major changes, like these, there then copy changes into the main overlay once behaviors have been confirmed.

hope this helps

**Comment by nabbi** [comment]
Thanks again for this PR. I did have to tweak the path references so it would properly source your 1.36.9999 instead of the default 9999. We might need to check the curl http response codes; observed 404 errors and a broken ebuild being pushed into the repo. More complexity now to account for :) 

commit d5f7bc8370f94544675b67ff55d612ab094df8e0

**Comment by dekeonus** [comment]
aah ... those 404s, In my testing the api.github.com service stopped talking to me, I thought it was a rate limiting behaviour as all URLs stopped not just the zoneminder. But perhaps api.github.com just has a coffee break every now and then.

I'll look into how to parse/handle curl response codes (it would be easiest if curl returns a non 0 error code so a _curl URL || {error msg and die}_ could be used)
As an alternative, adding checks at the end of vars_github_api() to ensure each of the variables holding a commit_sha_hash is exactly the length of a sha hash (i.e $tag_sha $commit_sha and iterate over the sub_module_shas).

**Comment by nabbi** [comment]
once again. nice work.

I move your functions into an include script to reduce our future touch points.
And added "logger" outputs to your new err_msg() as I run this headless in a crontab -- however I see now I have my "which" checks incorrect from an order of operations perspective :D I don't know which system doesn't have a logger command, yet will relook at this again and come up with a different approach.

7ca42f2e17b97667e49dfcdcbdab198edb7c2f4a

I update my server instance which runs this script daily. So this code is live from that regards.

**Comment by dekeonus** [comment]
hmm, I see the need for those error messages to go to stdout.
The reason I switched them all to stderr is to collect the error messages from curl_cmd and jq_cmd functions:
These need their stdout to be their parsed data.
As these both run in a subshell to collect their stdout, setting a variable for display later isn't available. They inherit their variables but can not export them back to the parent -this by design (of the shell).
I will ponder a solution (and I do note you can use some trickery with gdb but I consider that to be a *very bad idea*).

**Comment by nabbi** [comment]
I'm not overly concerned, yet I'll happily merge in any additional code you choose to develop. The main goal was to minimize my manual touch points and get zm point releases published shortly after they are released as upstream doesn't support the older builds.  There's at least [two other scripting approaches](https://github.com/search?q=ebuild+version+bump) of this workflow for inspiration. Next week (sorry) I can look at layering `repoman full -d` --  github actions run that upon commits and alerts me with issues -- this check would catch broken ebuilds, like the 404 fetches for instance, before it's pushed into the git repo. I typically run this, just wasn't in that script. Not sure if that changes your thought process.

**Comment by nabbi** [comment]
I incorporated repoman checks d88470cf61a7e99e926ab7307511476a7fc45583
This is probably super good enough. Feel free to reopen or PR if you have anything else outstanding. Thank again, this code is running and auto-updating the zoneminder ebuilds!
